bug: can't compile caltech-other/bool under macOS

Hi,

While compiling **cm3** under **macOS BigSur 11.6.1 x64** I encountered these bugs in **caltech-other/bool** C sources. 

All of them are comparisons between **signed** and **unsigned int**s.

```shell
cd /Users/brakmic/projects/downloads/cm3-dist-AMD64_LINUX-d5.11.4/caltech-other/bool
/usr/local/cm3/bin/cm3 -build -DBUILD_DIR=AMD64_DARWIN -DROOT=/Users/brakmic/projects/downloads/cm3-dist-AMD64_LINUX-d5.11.4 -DTARGET=AMD64_DARWIN -DM3_BACKEND_MODE=C
--- building in AMD64_DARWIN ---

ignoring ../src/m3overrides

[...snip...]
new source -> compiling bool.c
../src/bool.c: In function 'size_t hash(size_t, bool_t*, bool_t*)':
../src/bool.c:127:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]
  127 |     for (i=1; i < sizeof(void*); i++,c++)
      |               ~~^~~~~~~~~~~~~~~
../src/bool.c:130:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]
  130 |     for (i=0; i < sizeof(void*); i++,c++)
      |               ~~^~~~~~~~~~~~~~~
../src/bool.c:138:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]
  138 |     for (i=1; i < sizeof(void*); i++, c++)
      |               ~~^~~~~~~~~~~~~~~
../src/bool.c:141:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]
  141 |     for (i=0; i < sizeof(void*); i++, c++)
      |               ~~^~~~~~~~~~~~~~~
../src/bool.c:150:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]
  150 |     for (i=1; i < sizeof(void*); i++, c++)
      |               ~~^~~~~~~~~~~~~~~
../src/bool.c:153:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]
  153 |     for (i=0; i < sizeof(void*); i++, c++)
      |               ~~^~~~~~~~~~~~~~~
../src/bool.c:163:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]
  163 |     for (i=1; i < sizeof(void*); i++, c++)
      |               ~~^~~~~~~~~~~~~~~
../src/bool.c:166:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]
  166 |     for (i=0; i < sizeof(void*); i++, c++)
      |               ~~^~~~~~~~~~~~~~~
../src/bool.c: In function 'bool_t* bool_newvar(BOOL_T*)':
../src/bool.c:389:35: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]
  389 |     for (i=B->totvar-VAR_BLOCK; i < B->totvar; i++)
      |                                 ~~^~~~~~~~~~~
../src/bool.c: In function 'bool_t* _bool_substitute(BOOL_T*, bool_list_t*, bool_list_t*, bool_t*, int)':
../src/bool.c:964:14: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]
  964 |   while (loc < l1->n && b->id > l1->v[loc])
      |          ~~~~^~~~~~~
../src/bool.c:966:11: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]
  966 |   if (loc >= l1->n) {
      |       ~~~~^~~~~~~~
../src/bool.c:973:13: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]
  973 |     if (loc < l1->n && b->id == l1->v[loc]) {
      |         ~~~~^~~~~~~
../src/bool.c: In function 'bool_t* _bool_exists(BOOL_T*, bool_list_t*, bool_t*, int)':
../src/bool.c:1039:14: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]
 1039 |   while (loc < l1->n  && b->id > l1->v[loc])
      |          ~~~~^~~~~~~
../src/bool.c:1041:11: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]
 1041 |   if (loc >= l1->n) {
      |       ~~~~^~~~~~~~
cc1plus: all warnings being treated as errors
  compile_c => 1
C compiler failed compiling: ../src/bool.c
****  PARALLEL BACK-END BUILD, M3_PARALLEL_BACK = 20
new exporters -> recompiling BoolRemapImpl.i3
new exporters -> recompiling BoolImpl.i3
new exporters -> recompiling BoolVector.i3
new exporters -> recompiling BoolInteger.i3
new opaque info -> recompiling BoolIntegerImpl.m3
compilation failed => not building library "libbool.a"
Fatal Error: package build failed
Traceback (most recent call last):
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1816, in <module>
    Concierge().main()
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1750, in main
    command.execute()
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1502, in execute
    self.full_upgrade()
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1563, in full_upgrade
    self._upgrade.execute()
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1418, in execute
    self.buildship(self.packages())
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1066, in executor
    action.execute(paths)
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 839, in execute
    self.execute_path(package_path)
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 963, in execute_path
    self._buildglobal.execute_path(package_path)
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 913, in execute_path
    self.run(package_path, ["-build"])
  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 864, in run
    proc.check_returncode()
```

Ok, that's really pedantic.... Can't you just turn off that warning?

On Tue, Jan 18, 2022, 11:01 PM Harris Brakmić ***@***.***>
wrote:

> Hi,
>
> While compiling *cm3* under *macOS BigSur 11.6.1 x64* I encountered these
> bugs in *caltech-other/bool* C sources.
>
> All of them are comparisons between *signed* and *unsigned int*s.
>
> cd /Users/brakmic/projects/downloads/cm3-dist-AMD64_LINUX-d5.11.4/caltech-other/bool
> /usr/local/cm3/bin/cm3 -build -DBUILD_DIR=AMD64_DARWIN -DROOT=/Users/brakmic/projects/downloads/cm3-dist-AMD64_LINUX-d5.11.4 -DTARGET=AMD64_DARWIN -DM3_BACKEND_MODE=C
> --- building in AMD64_DARWIN ---
>
> ignoring ../src/m3overrides
>
> [...snip...]
> new source -> compiling bool.c
> ../src/bool.c: In function 'size_t hash(size_t, bool_t*, bool_t*)':../src/bool.c:127:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]  127 |     for (i=1; i < sizeof(void*); i++,c++)      |               ~~^~~~~~~~~~~~~~~../src/bool.c:130:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]  130 |     for (i=0; i < sizeof(void*); i++,c++)      |               ~~^~~~~~~~~~~~~~~../src/bool.c:138:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]  138 |     for (i=1; i < sizeof(void*); i++, c++)      |               ~~^~~~~~~~~~~~~~~../src/bool.c:141:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]  141 |     for (i=0; i < sizeof(void*); i++, c++)      |               ~~^~~~~~~~~~~~~~~../src/bool.c:150:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]  150 |     for (i=1; i < sizeof(void*); i++, c++)      |               ~~^~~~~~~~~~~~~~~../src/bool.c:153:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]  153 |     for (i=0; i < sizeof(void*); i++, c++)      |               ~~^~~~~~~~~~~~~~~../src/bool.c:163:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]  163 |     for (i=1; i < sizeof(void*); i++, c++)      |               ~~^~~~~~~~~~~~~~~../src/bool.c:166:17: error: comparison of integer expressions of different signedness: 'int' and 'long unsigned int' [-Werror=sign-compare]  166 |     for (i=0; i < sizeof(void*); i++, c++)      |               ~~^~~~~~~~~~~~~~~../src/bool.c: In function 'bool_t* bool_newvar(BOOL_T*)':../src/bool.c:389:35: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]  389 |     for (i=B->totvar-VAR_BLOCK; i < B->totvar; i++)      |                                 ~~^~~~~~~~~~~../src/bool.c: In function 'bool_t* _bool_substitute(BOOL_T*, bool_list_t*, bool_list_t*, bool_t*, int)':../src/bool.c:964:14: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]  964 |   while (loc < l1->n && b->id > l1->v[loc])      |          ~~~~^~~~~~~../src/bool.c:966:11: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]  966 |   if (loc >= l1->n) {      |       ~~~~^~~~~~~~../src/bool.c:973:13: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]  973 |     if (loc < l1->n && b->id == l1->v[loc]) {      |         ~~~~^~~~~~~../src/bool.c: In function 'bool_t* _bool_exists(BOOL_T*, bool_list_t*, bool_t*, int)':../src/bool.c:1039:14: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare] 1039 |   while (loc < l1->n  && b->id > l1->v[loc])      |          ~~~~^~~~~~~../src/bool.c:1041:11: error: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare] 1041 |   if (loc >= l1->n) {      |       ~~~~^~~~~~~~cc1plus: all warnings being treated as errors  compile_c => 1C compiler failed compiling: ../src/bool.c****  PARALLEL BACK-END BUILD, M3_PARALLEL_BACK = 20new exporters -> recompiling BoolRemapImpl.i3new exporters -> recompiling BoolImpl.i3new exporters -> recompiling BoolVector.i3new exporters -> recompiling BoolInteger.i3new opaque info -> recompiling BoolIntegerImpl.m3compilation failed => not building library "libbool.a"Fatal Error: package build failedTraceback (most recent call last):  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1816, in <module>    Concierge().main()  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1750, in main    command.execute()  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1502, in execute    self.full_upgrade()  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1563, in full_upgrade    self._upgrade.execute()  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1418, in execute    self.buildship(self.packages())  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 1066, in executor    action.execute(paths)  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 839, in execute    self.execute_path(package_path)  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 963, in execute_path    self._buildglobal.execute_path(package_path)  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 913, in execute_path    self.run(package_path, ["-build"])  File "/Users/brakmic/projects/downloads/build/../cm3-dist-AMD64_LINUX-d5.11.4/scripts/concierge.py", line 864, in run    proc.check_returncode()
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/modula3/cm3/issues/899>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKYNJMQGZYPRHSKH5QJSJDUWXPKBANCNFSM5MIJHH2A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/caltech-other/bool/src/bool.c
+++ b/caltech-other/bool/src/bool.c
@@ -116,7 +116,7 @@ static pairhash_t *hash_new (size_t sz)
 
 static size_t hash (size_t sz, bool_t *v1, bool_t *v2)
 {
-  int i;
+  unsigned int i;
   unsigned char h1, h2, h3, h4;
   size_t val;
   unsigned char *c;
@@ -383,7 +383,7 @@ bool_t *bool_newvar (BOOL_T *B)
     return NULL;
   }
   if (B->nvar == B->totvar) {
-    int i;
+    unsigned int i;
     B->totvar += VAR_BLOCK;
     REALLOC(B->H, pairhash_t *, B->totvar);
     for (i=B->totvar-VAR_BLOCK; i < B->totvar; i++)
@@ -956,7 +956,7 @@ bool_list_t *bool_qlist (BOOL_T *B, size_t n, bool_var_t *v)
 
 
 static bool_t *_bool_substitute (BOOL_T *B, bool_list_t *l1, bool_list_t *l2,
-				 bool_t *b, int loc)
+				 bool_t *b, unsigned int loc)
 {
   bool_t *b1;
   bool_t *l, *r;
@@ -1031,7 +1031,7 @@ bool_t *bool_substitute (BOOL_T *B, bool_list_t *l1, bool_list_t *l2,
 
 /*========================================================================*/  
 
-static bool_t *_bool_exists (BOOL_T *B, bool_list_t *l1, bool_t *b, int loc)
+static bool_t *_bool_exists (BOOL_T *B, bool_list_t *l1, bool_t *b, unsigned int loc)
 {
   bool_t *b1, *l, *r;
   static triplehash_t *tmp;

